actions/format_image_action.go

```diff
@@ -0,0 +1,157 @@
+
+Yaml syntax:
+ - action: format-image
+   imagename: image_name
+   imagesize: size
+   fs: filesystem
+   blocksize: 4096
+
+Mandatory properties:
+
+- imagename -- the name of the image file.
+
+- imagesize -- generated image size in human-readable form, examples: 100MB, 1GB, etc.
+
+- fs -- filesystem type used for formatting.
+
+Optional properties:
+
+- blocksize -- size of blocks in bytes, valid values are 1024, 2048 and 4096. For ext fs only.
+*/
+
+package actions
+
+import (
+	"fmt"
+	"github.com/docker/go-units"
+	"github.com/go-debos/fakemachine"
+	"os"
+	"path"
+	"strings"
+	"syscall"
+
+	"github.com/go-debos/debos"
+)
+
+type FormatImageAction struct {
+	debos.BaseAction `yaml:",inline"`
+	ImageName        string
+	ImageSize        string
+	FS               string
+	BlockSize        int
+	size             int64
+}
+
+func (i FormatImageAction) format(context debos.DebosContext) error {
+	label := fmt.Sprintf("Formatting image")
+
+	cmdline := []string{}
+	switch i.FS {
+	case "vfat":
+		cmdline = append(cmdline, "mkfs.vfat")
+	case "btrfs":
+		// Force formatting to prevent failure in case if partition was formatted already
+		cmdline = append(cmdline, "mkfs.btrfs", "-f")
+	case "none":
+	default:
+		cmdline = append(cmdline, fmt.Sprintf("mkfs.%s", i.FS))
+	}
+
+	if len(cmdline) != 0 {
+		cmdline = append(cmdline, context.Image)
+
+		cmd := debos.Command{}
+		if err := cmd.Run(label, cmdline...); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (i FormatImageAction) PreMachine(context *debos.DebosContext, m *fakemachine.Machine,
+	args *[]string) error {
+	image, err := m.CreateImage(i.ImageName, i.size)
+	if err != nil {
+		return err
+	}
+
+	context.Image = image
+	*args = append(*args, "--internal-image", image)
+	return nil
+}
+
+func (i *FormatImageAction) PreNoMachine(context *debos.DebosContext) error {
+	// XXXX Couldn't test this function.
+	//      I don't know which codepath leads here.
+	img, err := os.OpenFile(i.ImageName, os.O_WRONLY|os.O_CREATE, 0666)
+	if err != nil {
+		return fmt.Errorf("Couldn't open image file: %v", err)
+	}
+
+	err = img.Truncate(i.size)
+	if err != nil {
+		return fmt.Errorf("Couldn't resize image file: %v", err)
+	}
+
+	img.Close()
+
+	return nil
+}
+
+func (i *FormatImageAction) Run(context *debos.DebosContext) error {
+	i.LogStart()
+
+	err := i.format(*context)
+	if err != nil {
+		return err
+	}
+
+	context.ImageMntDir = path.Join(context.Scratchdir, "mnt")
+	os.MkdirAll(context.ImageMntDir, 0755)
+	err = syscall.Mount(context.Image, context.ImageMntDir, i.FS, 0, "")
+	if err != nil {
+		return fmt.Errorf("Mount failed: %v", err)
+	}
+
+	return nil
+}
+
+func (i FormatImageAction) Cleanup(context debos.DebosContext) error {
+	syscall.Unmount(context.ImageMntDir, 0)
+
+	return nil
+}
+
+func (i *FormatImageAction) Verify(context *debos.DebosContext) error {
+	// XXXX Should we really support 'fat32' as in image_partition_action.go?
```

Comment:
Wonder how much of this you could share with imagepartitoin

Reply:
Yep there's a good deal that can be factorized, but I'm not familiar with Go so I didn't give it a try yet. I can do that and issue a second PR (updating this one will be confusing if I move too much code).